test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs

```diff
@@ -27,6 +28,159 @@ public sealed partial class TestDataPipe : TestDataPipeBase
+                    "Sport\t2.71",
+                    "Technik & Wissen\t0.01"
+                });
+
+            // test key type with all invalid entries, and numeric type (should have missing value for every non-numeric label)
+            string name = TestName + "4-out.txt";
+            string pathOut = DeleteOutputPath("SavePipe", name);
+            using (var writer = OpenWriter(pathOut))
+            using (Env.RedirectChannelOutput(writer, writer))
+            {
+                TestCore(pathData, true,
+                    new[] {
+                        "loader=Text{col=RawLabel:TXT:0 col=Names:TXT:1-2 col=Features:TXT:3-4 header+}",
+                        string.Format("xf=TermLookup{{key=- col=FileLabelNum:RawLabel data={{{0}}}}}", mappingPathData),
+                        string.Format("xf=TermLookup{{col=FileLabelKey:RawLabel data={{{0}}}}}", mappingPathData),
+                        "xf=ChooseColumns{col=RawLabel col=FileLabelNum col=FileLabelKey}"
+                    }, suffix: "4");
+                writer.WriteLine(ProgressLogLine);
+                Env.PrintProgress();
+            }
+
+            CheckEqualityNormalized("SavePipe", name);
+
+            mappingPathData = DeleteOutputPath("SavePipe", "Mapping.txt");
+            File.WriteAllLines(mappingPathData,
+                new[] {
+                    "Wirtschaft\t10000000000",
+                    "Gesundheit\t10000000001",
+                    "Deutschland\t10000000002",
+                    "Ausland\t10000000003",
+                    "Unterhaltung\t10000000004",
+                    "Sport\t10000000005",
+                    "Technik & Wissen\t10000000006"
+                });
+
+            // test key type with all invalid entries, and numeric type (should have missing value for every non-numeric label)
+            TestCore(pathData, true,
+                new[] {
+                    "loader=Text{col=RawLabel:TXT:0 col=Names:TXT:1-2 col=Features:TXT:3-4 header+}",
+                    string.Format("xf=TermLookup{{col=FileLabel:RawLabel data={{{0}}}}}", mappingPathData),
+                    "xf=ChooseColumns{col=RawLabel col=FileLabel}"
+                }, suffix: "5");
+
+            Done();
+        }
+
+        [Fact]
+        public void SavePipeWithHeader()
+        {
+            string pathTerms = DeleteOutputPath("SavePipe", "Terms.txt");
+            File.WriteAllLines(pathTerms, new string[] {
+                "Amer-Indian-Eskimo",
+                "Black",
+                "Asian-Pac-Islander",
+                "White",
+            });
+
+            string pathData = GetDataPath("adult.train");
+            TestCore(pathData, false,
+                new[] {
+                    "loader=Text{header+ sep=comma col=Label:14 col=Age:0 col=Gender:TX:9 col=Mar:TX:5 col=Race:TX:8 col=Num:2,4,10-12 col=Txt:TX:~}",
+                    "xf=Cat{col=Race2:Key:Race data={" + pathTerms + "} termCol=Whatever}",
```

Comment:
>Race2:Key:Race [](start = 32, length = 14)

Is this supported?

Reply:
Are you referring to the "Race2:Key:Race" syntax? Or to the term data file argument? The first is supported, the second is supported in the latest iteration of this PR :-).

---
In reply to: [224882266](https://github.com/dotnet/machinelearning/pull/1236#discussion_r224882266) [](ancestors = 224882266)